src/lib/gadgets/sha256.ts

```diff
@@ -0,0 +1,306 @@
+// https://csrc.nist.gov/pubs/fips/180-4/upd1/final
+import { Field } from '../core.js';
+import { UInt32, UInt8 } from '../int.js';
+import { Bytes, FlexibleBytes } from '../provable-types/bytes.js';
+import { TupleN } from '../util/types.js';
+import { bitSlice, exists } from './common.js';
+import { Gadgets } from './gadgets.js';
+
+export { SHA256 };
+
+function padding(data: FlexibleBytes): UInt32[][] {
+  // pad the data with zeros to reach the desired length
+  // this is because we need to inherit to a static sized Bytes array. unrelated to sha256
+  let zeroPadded = Bytes.from(data);
+
+  // now pad the data to reach the format expected by sha256
+  // pad 1 bit, followed by k zero bits where k is the smallest non-negative solution to
+  // l + 1 + k = 448 mod 512
+  // then append a 64bit block containing the length of the original message in bits
+
+  let l = zeroPadded.length * 8; // length in bits
+  let k = (448 - (l + 1)) % 512;
+
+  // pad message exceeds 512bit and needs a new block
+  if (k < 0) k += 512;
+
+  let paddingBits = (
+    '1' + // append 1 bit
+    '0'.repeat(k) + // append k zero bits
+    '0'.repeat(64 - l.toString(2).length) + // append 64bit containing the length of the original message
+    l.toString(2)
+  ).match(/.{1,8}/g)!; // this should always be devisable by 8
```

Comment:
woah what kind of black magic is this 🥷
we also have a `chunkString()` helper but please leave this in for its educational value :D